### ISSUE TITLE
Cancel pending request like google map

### ISSUE THREAD

**Issue by jaroslawwroblewski** [openseadragon/openseadragon#1747]
Hello so far I funded on options "imageLoaderLimit" (number of concurrent image requests) but I want resolve my problem like google team - by  cancel any unnecessary request (when I change zoom or position). How can I do that?

**Comment by iangilman** [comment]
Once a tile request has started communicating with the server, there's currently no way for you to abort it. To make that happen, we'd need to add the functionality somehow to the ImageLoader. What sort of API would be good? Would you want to abort all of the in flight requests, or would you want to look through them and pick and choose?

Would you be up for working on a patch for this new feature?

Meanwhile, the imageLoaderLimit is your best bet... for instance, if you keep it to `1`, then there will only be one load in flight at any time. Once it completes, it'll load whatever is the most important tile at that moment, so you know you're always getting the best tiles for the view.

**Comment by jaroslawwroblewski** [comment]
> Would you want to abort all of the in flight requests, or would you want to look through them and pick and choose?

I would like to cancel all that are still saving (pending) retaining images that have already downloaded

> Would you be up for working on a patch for this new feature?

sure, we just need to clarify what we want to achieve so that two sides can benefit from it :)

**Comment by xgui3783** [comment]
I believe if it was an XHR request, then there is an [abort](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/abort) method. 

Digging a little into the code, it seems the relevant part of the code in OSD <https://github.com/openseadragon/openseadragon/blob/c01887f/src/openseadragon.js#L2158-L2191>, the xhr request object can either be an xhr request, or an activexobject, if the user tried to access `file://`. 

AFAIK, activexobject <https://developer.mozilla.org/en-US/docs/Archive/Web/JavaScript/Microsoft_Extensions/ActiveXObject> does not seem to support abort functionality.

I am actually interested to implement a patch, if it is a feature that you would like in OSD.

**Comment by iangilman** [comment]
Yes, this functionality would be welcome! I'm still not really sure exactly what API would be best, but I suppose the first step is having it available at a low level.

Note that most tile loads happen via the browser's image loading mechanism, not through XHR (though XHR is an option). It would be good to provide a way to abort those as well. Looks like it's possible, according to this post:

https://stackoverflow.com/a/5278475/306351

… simply by replacing the `src` attribute.

Anyway, @xgui3783, @jaroslawwroblewski please confer here on the proposed patch so we are all on the same page, and everyone knows who is doing what :-)

**Comment by xgui3783** [comment]
@iangilman aww, I had not noticed that most of the images load via `img.src`.

my concern of setting `src` to an empty string is that since it is a non standard behaviour, it could potentially break in the future, should the browser decide to patch it.

Am I correct in assuming that the reason XHR requests are not used in all circumstances is
- backwards compat or
- circumvent cors

?

**Comment by iangilman** [comment]
Well, OpenSeadragon has been around for quite a while; CORS wasn't really a thing back when OSD got started. So the answer to "why doesn't it default to XHR" is partially just historical. We even have a JSONP loading option for tile sources, if that tells you anything!

That said, I don't know what the implications of switching over to XHR are at this point. I don't think I'd want to make it the default, because it would likely break at least some existing OpenSeadragon uses, and at any rate it adds one more layer of complication to people getting set up in the first place (having to make sure their server is properly configured). Images have the advantage of working without extra configuration in most scenarios.

Anyway, I think the cancellation feature would still be useful, even if it's only for the XHR path… People who want cancellation can go the extra mile to make sure they're set up for XHR.

But also, it doesn't necessarily seem too risky to use the empty string solution for image loading… If the browsers stop supporting it, I would think it just wouldn't abort, but there wouldn't necessarily be other negative side effects. Are there possible effects I'm not considering?
